Framework/DataObjects/src/EventList.cpp

```diff
@@ -1849,6 +1848,70 @@ void EventList::histogramForWeightsHelper(const std::vector<T> &events, const Ma
   std::transform(E.cbegin(), E.cend(), E.begin(), static_cast<double (*)(double)>(sqrt));
 }
 
+// --------------------------------------------------------------------------
+/** Generates both the Y and E (error) histograms
+ * for an EventList with WeightedEvents.
+ *
+ * This histograms without sorting the events first by using the bin step size to estimate the bin number. This only
+ * works for logarithmic or linear binning.
+ *
+ * @param events: vector of events (with weights)
+ * @param step: bin step size
+ * @param X: X-bins supplied
+ * @param Y: counts returned
+ * @param E: errors returned
+ * @throw runtime_error if the EventList does not have weighted events
+ */
+template <class T>
+void EventList::histogramForWeightsHelper(const std::vector<T> &events, const double step, const MantidVec &X,
+                                          MantidVec &Y, MantidVec &E) {
+  size_t x_size = X.size();
+
+  if (x_size <= 1) {
+    // X was not set. Return an empty array.
+    Y.resize(0, 0);
+    return;
+  }
+
+  Y.resize(x_size - 1, 0.0);
+  E.resize(x_size - 1, 0.0);
+
+  if (events.empty())
+    return;
+
+  const auto xmin = X.front();
+  const auto xmax = X.back();
+
+  double divisor, offset;
+  boost::optional<size_t> (*findBin)(const MantidVec &, const double, const double, const double);
```

Comment:
A comment here about this being a function pointer would be helpful. This might be clearer if it were switched to [std::function](https://en.cppreference.com/w/cpp/utility/functional/function).

Reply:
I tried changing this to `std::function<boost::optional<size_t>(const MantidVec &, const double, const double, const double)> findBin;` but it slowed it down by 10-20%